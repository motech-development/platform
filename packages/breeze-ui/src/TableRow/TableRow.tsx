import React, { FC, HTMLAttributes, memo } from 'react';
import { darken } from 'polished';
import styled, { ThemeProvider } from 'styled-components';

const tableRowTheme = {
  default: {
    background: '#fff',
    border: '#eee',
    colour: '#000',
  },
  primary: {
    background: '#007fa8',
    border: darken(0.02, '#007fa8'),
    colour: '#fff',
  },
  secondary: {
    background: '#f6f9fc',
    border: '#eee',
    colour: '#333',
  },
};

interface IBaseTableRow {
  $colour: keyof typeof tableRowTheme;
}

const BaseTableRow = styled.tr<IBaseTableRow>`
  ${({ $colour, theme }) => `
    background-color: ${theme[$colour].background};
    border-bottom: 2px solid ${theme[$colour].border};
    color: ${theme[$colour].colour};
  `}
`;

export interface ITableRowProps extends HTMLAttributes<HTMLTableRowElement> {
  colour?: keyof typeof tableRowTheme;
}

const TableRow: FC<ITableRowProps> = ({ colour = 'default', ...rest }) => (
  <ThemeProvider theme={tableRowTheme}>
    {/* eslint-disable-next-line react/jsx-props-no-spreading */}
    <BaseTableRow $colour={colour} {...rest} />
  </ThemeProvider>
);

export default memo(TableRow);
