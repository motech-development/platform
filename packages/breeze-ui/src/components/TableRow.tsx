import { FC, forwardRef, TableHTMLAttributes } from 'react';
import { classNames, themeClass } from '../utils/className';
import TTheme from '../utils/theme';

export interface ITableRowProps
  extends TableHTMLAttributes<HTMLTableRowElement> {
  colour?: TTheme;
}

const TableRow: FC<ITableRowProps> = forwardRef<
  HTMLTableRowElement,
  ITableRowProps
>(({ className = '', colour = 'secondary', ...rest }, ref) => (
  <>
    {/* @tailwind: bg-blue-700 bg-gray-100 bg-green-700 bg-red-700 bg-yellow-700 */}
    {/* @tailwind: border-blue-800 border-gray-200 border-green-800 border-red-800 border-yellow-800 */}
    {/* @tailwind: text-blue-100 text-gray-700 text-green-100 text-red-100 text-yellow-100 */}
    <tr
      className={classNames(
        'border-b-2',
        className,
        themeClass(
          colour,
          'bg-{theme}-700 border-{theme}-800 text-{theme}-100',
          {
            secondary: 'bg-{theme}-100 border-{theme}-200 text-{theme}-700',
          },
        ),
      )}
      ref={ref}
      {...rest}
    />
  </>
));

export default TableRow;
