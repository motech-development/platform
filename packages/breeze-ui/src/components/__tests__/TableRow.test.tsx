import { render } from '@testing-library/react';
import TableRow from '../TableRow';

describe('TableRow', () => {
  it('should have the correct theme by default', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow>
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the correct primary theme styles', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow colour="primary">
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the correct secondary theme styles', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow colour="secondary">
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the correct success theme styles', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow colour="success">
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the correct warning theme styles', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow colour="warning">
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should have the correct danger theme styles', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow colour="danger">
            <td>Test</td>
          </TableRow>
        </tbody>
      </table>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
