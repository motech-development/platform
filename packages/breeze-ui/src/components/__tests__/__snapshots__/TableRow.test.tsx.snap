// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should have the correct danger theme styles 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-red-700 border-red-800 text-red-100"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the correct primary theme styles 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the correct secondary theme styles 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the correct success theme styles 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-green-700 border-green-800 text-green-100"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the correct theme by default 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the correct warning theme styles 1`] = `
<table>
  <tbody>
    <tr
      class="border-b-2 bg-yellow-700 border-yellow-800 text-yellow-100"
    >
      <td>
        Test
      </td>
    </tr>
  </tbody>
</table>
`;
